File: plat/ts/entities/daikon.ts
import * as Assets from 'assets';

import Entity from '../entity.js';
import Dir from '../dir.js';
import { Player } from './player.js';
import { World } from '../game.js';
import { FSMEntity, FSMStates } from '../fsmentity.js';
import Phys from '../phys.js';
import CollisionType from '../collisiontype.js';

enum Frames {
  Idle = 0,
  Rise,
  Fall,
  Pain
}

enum States {
  None,
  Wait,
  Pop,
  Recharge
}

class Daikon extends FSMEntity {
  type = 'Daikon';
  sprite = Assets.find('daikon');
  drawOfs: [number, number] = [-2, -3];

  state = States.Wait;
  nextState = States.None;
  nextStateTime = 0;

  die() {
    super.die();
    World().spawnDeathParticle(this, Frames.Pain);
  }

  #states: FSMStates = {
    default: {
      enter: () => this.fsmDefaultTransitionTo(States.Wait),
    },

    [States.Wait]: {
      enter: () => this.frame = Frames.Idle,
      update: (ticks: number) => {
        if (Math.abs(World().player.center(0) - this.center(0)) < 40) {
          this.fsmTransitionTo(States.Pop);
        }
      },
      canCollide: (other: Entity, dir: Dir) => CollisionType.Disabled,
    },

    [States.Pop]: {
      enter: () => {
        this.vel[1] = -3;
        this.frame = Frames.Rise;
      },
      update: (ticks: number) => {
        this.vel[1] += Phys.enemyGravity;
        if (!this.moveY(this.vel[1])) {
          this.fsmTransitionTo(States.Recharge);
        }
      },
      canCollide: (other: Entity, dir: Dir) => {
        if (other instanceof Player && dir == Dir.Up) return CollisionType.Platform;
        else if (other instanceof Player) return CollisionType.Trigger;
        else return CollisionType.Disabled;
      }
    },

    [States.Recharge]: {
      enter: () => {
        this.fsmTransitionAtTime(States.Wait, 120);
        this.frame = Frames.Idle;
      },
      canCollide: (other: Entity, dir: Dir) => CollisionType.Disabled,
    }
  }

  update(ticks: number, dt: number) {
    this.fsmUpdate(this.#states, ticks);
  }

  canCollide(other: Entity, dir: Dir) {
    return this.fsmCanCollide(this.#states, other, dir);
  }

  collide(other: Entity, dir: Dir) {
    this.handlePlayerStomp(other, dir);
  }
}

export { Daikon };

File: csgame/entities/Daikon.cs
namespace Daikon;
using Slate2D;

enum Frames
{
    Idle,
    Rise,
    Fall,
    Pain
}

enum States
{
    Default,
    Wait,
    Pop,
    Recharge
}

[Spawnable]
class Daikon : FSMEntity<States>
{
    public Daikon(LDTKEntity ent) : base(ent)
    {
        Sprite = Assets.Find("daikon");
        DrawOfs = (-2, -3);
        FSMTransitionTo(States.Wait);
    }

    public override void Die()
    {
        base.Die();
        Main.World.SpawnDeathParticle(this, (uint)Frames.Pain);
    }

    void Default_Collide(Entity other, Dir dir)
    {
        HandlePlayerStomp(other, dir);
        if (other is WorldEnt) {
            FSMTransitionTo(States.Recharge);
        }
    }

    void Wait_Enter() => Frame = (uint)Frames.Idle;
    void Wait_Update(uint ticks, float dt)
    {
        if (Math.Abs(Main.World.Player.Center.X - Center.X) < 40)
        {
            FSMTransitionTo(States.Pop);
        }
    }
    CollisionType Wait_CanCollide(Entity other, Dir dir) => CollisionType.Disabled;

    void Pop_Enter()
    {
        Vel.Y = -3;
        Frame = (uint)Frames.Rise;
    }
    void Pop_Update(uint ticks, float dt)
    {
        Vel.Y += Phys.EnemyGravity;
        MoveY(Vel.Y);
    }
    CollisionType Pop_CanCollide(Entity other, Dir dir)
    {
        if (other is Player && dir == Dir.Up) return CollisionType.Platform;
        else if (other is Player) return CollisionType.Trigger;
        else return CollisionType.Disabled;
    }

    void Recharge_Enter()
    {
        FSMTimer(States.Wait, 120);
        Frame = (uint)Frames.Idle;
    }
    void Recharge_Collide(Entity other, Dir dir)
    {
        HandlePlayerStomp(other, dir);
        if (other is WorldEnt) FSMTransitionTo(States.Recharge);
    }
}